nx/source/services/nfc.c

```diff
@@ -0,0 +1,819 @@
+    return _nfpuInterfaceCmdInIdOutEvent(&g_nfpuInterface, 18, id, out);
+}
+
+Result nfpuAttachAvailabilityChangeEvent(Event *out) {
+    IpcCommand c;
+    ipcInitialize(&c);
+
+    struct {
+        u64 magic;
+        u64 cmd_id;
+    } *raw;
+
+    raw = serviceIpcPrepareHeader(&g_nfpuInterface, &c, sizeof(*raw));
+
+    raw->magic = SFCI_MAGIC;
+    raw->cmd_id = 23;
+
+    Result rc = serviceIpcDispatch(&g_nfpuInterface);
+
+    if (R_SUCCEEDED(rc)) {
+        IpcParsedCommand r;
+        struct {
+            u64 magic;
+            u64 result;
+        } *resp;
+
+        serviceIpcParse(&g_nfpuInterface, &r, sizeof(*resp));
+        resp = r.Raw;
+
+        rc = resp->result;
+
+        if (R_SUCCEEDED(rc))
+            eventLoadRemote(out, r.Handles[0], false);
+    }
+
+    return rc;
+}
+
+Result nfpuGetState(NfpuState *out) {
+    IpcCommand c;
+    ipcInitialize(&c);
+
+    struct {
+        u64 magic;
+        u64 cmd_id;
+    } *raw;
+
+    raw = serviceIpcPrepareHeader(&g_nfpuInterface, &c, sizeof(*raw));
+
+    raw->magic = SFCI_MAGIC;
+    raw->cmd_id = 19;
+
+    Result rc = serviceIpcDispatch(&g_nfpuInterface);
+
+    if (R_SUCCEEDED(rc)) {
+        IpcParsedCommand r;
+        struct {
+            u64 magic;
+            u64 result;
+            u32 state;
+        } *resp;
+
+        serviceIpcParse(&g_nfpuInterface, &r, sizeof(*resp));
+        resp = r.Raw;
+
+        rc = resp->result;
+
+        if (R_SUCCEEDED(rc) && out)
+            *out = resp->state;            
+    }
+
+    return rc;
+}
+
+Result nfpuGetDeviceState(HidControllerID id, NfpuDeviceState *out) {
+    if (id == CONTROLLER_P1_AUTO)
+        return nfpuGetDeviceState(g_controllerP1AutoID, out);
+
+    IpcCommand c;
+    ipcInitialize(&c);
+
+    struct {
+        u64 magic;
+        u64 cmd_id;
+        u64 id;
+    } *raw;
+
+    raw = serviceIpcPrepareHeader(&g_nfpuInterface, &c, sizeof(*raw));
+
+    raw->magic = SFCI_MAGIC;
+    raw->cmd_id = 20;
+    raw->id = hidControllerIDToOfficial(id);
+
+    Result rc = serviceIpcDispatch(&g_nfpuInterface);
+
+    if (R_SUCCEEDED(rc)) {
+        IpcParsedCommand r;
+        struct {
+            u64 magic;
+            u64 result;
+            u32 state;
+        } *resp;
+
+        serviceIpcParse(&g_nfpuInterface, &r, sizeof(*resp));
+        resp = r.Raw;
+
+        rc = resp->result;
+
+        if (R_SUCCEEDED(rc) && out)
+            *out = resp->state;            
+    }
+
+    return rc;
+}
+
+Result nfpuListDevices(u32 *count, HidControllerID *out, size_t num_elements) {
+    // This is the maximum number of controllers that can be connected to a console at a time
+    // Incidentally, this is the biggest value official software (SSBU) was observed using
+    #define MAX_CONTROLLERS 9
+
+    IpcCommand c;
+    ipcInitialize(&c);
+
+    u64 buf[MAX_CONTROLLERS] = {0};
+    ipcAddRecvStatic(&c, buf, MAX_CONTROLLERS * sizeof(u64), 0);
```

Comment:
Add: "if (num_elements > MAX_CONTROLLERS) num_elements = MAX_CONTROLLERS;" Then use "i<num_elements" with the for-loop.

Also replace the define with a stack-var.